homeassistant/components/feedreader.py

```diff
@@ -0,0 +1,76 @@
+"""RSS/Atom feed reader for Home Assistant."""
+from datetime import datetime
+from logging import getLogger
+from homeassistant.helpers import validate_config
+from homeassistant.helpers.event import track_utc_time_change
+
+
+REQUIREMENTS = ['feedparser==5.2.1']
+
+_LOGGER = getLogger(__name__)
+
+DOMAIN = "feedreader"
+EVENT_FEEDREADER = "feedreader"
+
+
+# pylint: disable=too-few-public-methods
+class FeedManager(object):
+    """Abstraction over feedparser module."""
+
+    def __init__(self, url, hass):
+        """Initialize the FeedManager object, poll every hour."""
+        self._url = url
+        self._feed = None
+        self._hass = hass
+        # Initialize last entry timestamp as epoch time
+        self._last_entry_timestamp = datetime.utcfromtimestamp(0).timetuple()
+        _LOGGER.debug('Loading feed %s', self._url)
+        track_utc_time_change(hass, lambda now: self._update(),
+                              minute=0, second=0)
+
+    def _log_no_entries(self):
+        """Send no entries log at debug level."""
+        _LOGGER.debug('No new entries in feed %s', self._url)
+
+    def _update(self):
+        """Update the feed and publish new entries in the event bus."""
+        import feedparser
+        _LOGGER.info('Fetching new data from feed %s', self._url)
+        self._feed = feedparser.parse(self._url,
+                                      etag=None if not self._feed
+                                      else self._feed.get('etag'),
+                                      modified=None if not self._feed
+                                      else self._feed.get('modified'))
+        if self._feed and self._feed.bozo != 0:
+            _LOGGER.error('Error parsing feed %s', self._url)
+        # Using etag and modified, if there's no new data available,
+        # the entries list will be empty
+        elif len(self._feed.entries) > 0:
+            self._publish_new_entries()
+            self._last_entry_timestamp = self._feed.entries[0].published_parsed
+        else:
+            self._log_no_entries()
+
+    def _publish_new_entries(self):
+        """Publish new entries to the event bus."""
+        new_entries = False
+        for entry in self._feed.entries:
+            # Consider only entries newer then the latest parsed one
+            if entry.published_parsed > self._last_entry_timestamp:
+                new_entries = True
+                entry.update({'feed_url': self._url})
+                self._hass.bus.fire(EVENT_FEEDREADER, entry)
+        if not new_entries:
+            self._log_no_entries()
+
+
+def setup(hass, config):
+    """Setup the feedreader component."""
+    if not validate_config(config, {DOMAIN: ['urls']}, _LOGGER):
```

Comment:
Sorry, make it Url() in config schema

On Sun, Apr 17, 2016, 16:22 Alexander Fortin <notifications@github.com>
wrote:

> In homeassistant/components/feedreader.py
> <https://github.com/home-assistant/home-assistant/pull/1836#discussion_r59995230>
> :
>
> > +    def _publish_new_entries(self):
> > +        """Publish new entries to the event bus."""
> > +        new_entries = False
> > +        for entry in self._feed.entries:
> > +            # Consider only entries newer then the latest parsed one
> > +            if entry.published_parsed > self._last_entry_timestamp:
> > +                new_entries = True
> > +                entry.update({'feed_url': self._url})
> > +                self._hass.bus.fire(EVENT_FEEDREADER, entry)
> > +        if not new_entries:
> > +            self._log_no_entries()
> > +
> > +
> > +def setup(hass, config):
> > +    """Setup the feedreader component."""
> > +    if not validate_config(config, {DOMAIN: ['urls']}, _LOGGER):
>
> I did but config.get() now returns functions and not strings, e.g.
>
> 16-04-17 23:19:39 DEBUG (MainThread) [homeassistant.components.feedreader]
> Loading feed <function Url at 0x7f3934f4af28>
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/1836/files/197230c8f0cd819278379a008fd25700d4938319#r59995230>
>